core/embed/extmod/modtrezorui/display.h

```diff
@@ -61,9 +73,21 @@ void display_clear(void);
 void display_bar(int x, int y, int w, int h, uint16_t c);
 void display_bar_radius(int x, int y, int w, int h, uint16_t c, uint16_t b,
                         uint8_t r);
+void display_bar_radius_buffer(int x, int y, int w, int h, uint8_t r,
+                               buffer_text_t *buffer);
 
 bool display_toif_info(const uint8_t *buf, uint32_t len, uint16_t *out_w,
                        uint16_t *out_h, toif_format_t *out_format);
+
+bool display_jpeg_info(const uint8_t *data, uint32_t len, uint16_t *out_w,
+                       uint16_t *out_h, uint16_t *out_mcu_height);
+void display_jpeg(int x, int y, const uint8_t *data, uint32_t datalen);
+
+void display_jpeg_buffer_prepare(JDEC *jd, jpeg_context_t *jpg_context,
+                                 uint8_t *buffer, const uint8_t *data,
+                                 uint32_t datalen, uint32_t line_width);
+void display_jpeg_buffer_decomp(JDEC *jd);
```

Comment:
Good idea. Rewritten to rust: https://github.com/trezor/trezor-firmware/commit/1fe1216f5f4c0a7520e843da758803aa22e2fe48 (and removed some leftovers https://github.com/trezor/trezor-firmware/commit/11f208e62522e7ca942e4f2e26326dde339394d6).

Moved the jpeg work data to buffers, but i am not happy about what became of that functionality. Originally i needed that because the buffers for rendering need to be placed in memory in a way accessible by the dma2d, so it needs special care in bootloader. Some of the buffers don't care and could be placed anywhere, they just don't fit on stack. Maybe the buffers for jpeg and blurring could be stored in heap/GC, as we most likely won't be using this in bootloader. 

Dma2d in also not used in bootloader yet, but that may change later.